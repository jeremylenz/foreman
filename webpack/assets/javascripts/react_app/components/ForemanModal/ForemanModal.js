import React from 'react';
import { Modal } from 'patternfly-react';
import PropTypes from 'prop-types';
import ModalContext from './ForemanModalContext';
import { extractModalNodes } from './helpers';

const ForemanModal = props => {
  const { id, title, onClose, openState, children, ...propsToPassDown } = props;
  // Extract header and footer from children, if provided
  const { headerChild, footerChild, otherChildren } = extractModalNodes(
    children
  );
  const isOpen = false || (openState && openState.open);
  const context = {
    isOpen,
    onClose,
    title,
  };

  return (
    <ModalContext.Provider value={context}>
      {/* Change the name of the props we are passing down to Modal to conform to Patternfly 3 api */}
      <Modal
        onHide={onClose}
        show={isOpen}
        className="foreman-modal"
        {...propsToPassDown}
      >
        {headerChild}
        <Modal.Body>{otherChildren}</Modal.Body>
        {footerChild}
      </Modal>
    </ModalContext.Provider>
  );
};

ForemanModal.propTypes = {
  children: PropTypes.node,
  title: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  openState: PropTypes.shape({
    open: PropTypes.bool,
  }),
  onClose: PropTypes.func.isRequired,
};

ForemanModal.defaultProps = {
  children: null,
  openState: {
    open: false,
  },
};

export default ForemanModal;
