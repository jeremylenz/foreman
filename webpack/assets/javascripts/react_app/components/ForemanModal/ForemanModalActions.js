import {
  ADD_MODAL,
  SET_MODAL_OPEN,
  SET_MODAL_CLOSED,
} from './ForemanModalConstants';

// In function components, no need to use these-- just use useDispatch and dispatch the actions directly.
// For class components, these are provided so you can plug them into bindActionCreators

export const addModal = ({ id, open = false }) => (dispatch, getState) => {
  const { foremanModals } = getState();
  if (foremanModals && foremanModals[id]) {
    throw new Error(`ForemanModal with ID ${id} already exists`);
  }
  return dispatch({
    type: ADD_MODAL,
    payload: { id, open },
  });
};

export const setModalOpen = ({ id }) => (dispatch, getState) => {
  const { foremanModals } = getState();
  if (foremanModals && !foremanModals[id]) {
    throw new Error(
      `SET_MODAL_OPEN error: Modal with id '${id}' does not exist`
    );
  }
  return dispatch({
    type: SET_MODAL_OPEN,
    payload: { id },
  });
};

export const setModalClosed = ({ id }) => (dispatch, getState) => {
  const { foremanModals } = getState();
  if (foremanModals && !foremanModals[id]) {
    throw new Error(
      `SET_MODAL_CLOSED error: Modal with id '${id}' does not exist`
    );
  }
  return dispatch({
    type: SET_MODAL_CLOSED,
    payload: { id },
  });
};

// Pass in the ForemanModal id here and get bound action creators with the id already plugged in.
export const bindForemanModalActionsToId = ({ id }) => ({
  addModal: () => addModal({ id }),
  setModalOpen: () => setModalOpen({ id }),
  setModalClosed: () => setModalClosed({ id }),
});
