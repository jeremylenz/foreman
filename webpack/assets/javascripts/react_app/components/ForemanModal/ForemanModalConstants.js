export const SET_MODAL_OPEN = 'SET_MODAL_OPEN';
export const SET_MODAL_CLOSED = 'SET_MODAL_CLOSED';
export const ADD_MODAL = 'ADD_MODAL';
