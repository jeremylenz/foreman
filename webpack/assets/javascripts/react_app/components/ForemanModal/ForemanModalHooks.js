import { useContext } from 'react';
import ModalContext from './ForemanModalContext';

// Because enzyme doesn't support useContext yet
export const useModalContext = () => useContext(ModalContext);

// to get enzyme hacky test to work
export default ModalContext;
