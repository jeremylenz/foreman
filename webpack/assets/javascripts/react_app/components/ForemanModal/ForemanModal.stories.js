import React from 'react';
import { Button } from 'patternfly-react';
import { storiesOf } from '@storybook/react';
import storeDecorator from '../../../../../stories/storeDecorator';
import ForemanModal from '.';
import { useForemanModal } from './ForemanModalHooks';
import Story from '../../../../../stories/components/Story';

storiesOf('Components/ForemanModal', module)
  .addDecorator(storeDecorator) // add Redux store to story
  .add('With default header & footer', () =>
    // using createElement here so that hooks work in stories
    React.createElement(() => {
      const [, toggleModal] = useForemanModal({ id: 'default' }); // not using modalOpen so not assigning it
      return (
        <Story>
          <Button bsStyle="primary" onClick={toggleModal}>
            Show Modal
          </Button>
          <ForemanModal id="default" title="I'm a modal!">
            <ForemanModal.Header />
            This is the modal body
            <ForemanModal.Footer />
          </ForemanModal>
        </Story>
      );
    })
  );

storiesOf('Components/ForemanModal', module)
  .addDecorator(storeDecorator) // add Redux store to story
  .add('With custom header & footer', () =>
    React.createElement(() => {
      const [, toggleModal] = useForemanModal({ id: 'custom' });
      return (
        <Story>
          <Button bsStyle="primary" onClick={toggleModal}>
            Show Modal
          </Button>
          <ForemanModal id="custom" title="I'm a custom modal!">
            <ForemanModal.Header>
              <h3>This is a custom header! :)</h3>
            </ForemanModal.Header>
            If a {`<ForemanModal.Header>`} is provided AND has children, the
            title prop will be ignored.
            <ForemanModal.Footer>
              Click the X in the upper right to close
            </ForemanModal.Footer>
          </ForemanModal>
        </Story>
      );
    })
  );

storiesOf('Components/ForemanModal', module)
  .addDecorator(storeDecorator) // add Redux store to story
  .add('With unordered header & footer', () =>
    React.createElement(() => {
      const [, toggleModal] = useForemanModal({ id: 'custom' });
      return (
        <Story>
          <Button bsStyle="primary" onClick={toggleModal}>
            Show Modal
          </Button>
          <ForemanModal id="custom" title="I'm a custom modal!">
            <div>
              Header and footer will be correctly ordered when rendering, even
              if they are out of order in the markup
            </div>
            <ForemanModal.Footer>This is the footer</ForemanModal.Footer>
            <ForemanModal.Header>
              <h3>This is the header</h3>
            </ForemanModal.Header>
          </ForemanModal>
        </Story>
      );
    })
  );

storiesOf('Components/ForemanModal', module)
  .addDecorator(storeDecorator) // add Redux store to story
  .add('With no close button', () =>
    React.createElement(() => {
      const [, toggleModal] = useForemanModal({ id: 'noClose' });
      return (
        <Story>
          <Button bsStyle="primary" onClick={toggleModal}>
            Show Modal
          </Button>
          <ForemanModal id="noClose" title="I'm an X-less modal!">
            <ForemanModal.Header closeButton={false} />
            <br />
            Props passed to ForemanModal.Header will be passed down to
            Modal.Header
            <ForemanModal.Footer />
          </ForemanModal>
        </Story>
      );
    })
  );

storiesOf('Components/ForemanModal', module)
  .addDecorator(storeDecorator) // add Redux store to story
  .add('With no footer', () =>
    React.createElement(() => {
      const [, toggleModal] = useForemanModal({ id: 'noFooter' });
      return (
        <Story>
          <Button bsStyle="primary" onClick={toggleModal}>
            Show Modal
          </Button>
          <ForemanModal id="noFooter" title="I'm a modal!">
            <ForemanModal.Header />
            This is the modal body. There is no footer.
          </ForemanModal>
        </Story>
      );
    })
  );

storiesOf('Components/ForemanModal', module)
  .addDecorator(storeDecorator) // add Redux store to story
  .add('With no header', () =>
    React.createElement(() => {
      const [, toggleModal] = useForemanModal({ id: 'noHeader' });
      return (
        <Story>
          <Button bsStyle="primary" onClick={toggleModal}>
            Show Modal
          </Button>
          <ForemanModal id="noHeader" title="I'm a modal!">
            This is the modal body. There is no header.
            <ForemanModal.Footer />
          </ForemanModal>
        </Story>
      );
    })
  );

storiesOf('Components/ForemanModal', module)
  .addDecorator(storeDecorator) // add Redux store to story
  .add('With props passed down via spread syntax', () =>
    React.createElement(() => {
      const [, toggleModal] = useForemanModal({ id: 'propsPassed' });
      return (
        <Story>
          <Button bsStyle="primary" onClick={toggleModal}>
            Show Modal
          </Button>
          <ForemanModal id="propsPassed" title="I'm a modal!" myProp="Hii">
            <ForemanModal.Header />
            The inner {`<Modal>`} component will have any props you pass to
            {`<ForemanModal>`}. (Look in the React dev tools for
            &lsquo;myProp&rsquo;)
            <ForemanModal.Footer />
          </ForemanModal>
        </Story>
      );
    })
  );
