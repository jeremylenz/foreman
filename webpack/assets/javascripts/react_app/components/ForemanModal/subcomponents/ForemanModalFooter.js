import React from 'react';
import PropTypes from 'prop-types';
import { Modal, Button } from 'patternfly-react';
import { useModalContext } from '../ForemanModalHooks';

const ForemanModalFooter = props => {
  const childCount = React.Children.count(props.children);
  const { onClose } = useModalContext();
  // Render the provided children, or default markup if none given
  return (
    <Modal.Footer {...props}>
      {props.children}
      {childCount === 0 && (
        <Button bsStyle="default" onClick={onClose}>
          Close
        </Button>
      )}
    </Modal.Footer>
  );
};

ForemanModalFooter.propTypes = {
  children: PropTypes.node,
};

ForemanModalFooter.defaultProps = {
  children: null,
};

export default ForemanModalFooter;
