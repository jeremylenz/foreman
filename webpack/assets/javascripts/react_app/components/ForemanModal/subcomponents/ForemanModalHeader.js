import React from 'react';
import PropTypes from 'prop-types';
import { Modal } from 'patternfly-react';
import { useModalContext } from '../ForemanModalHooks';

const ForemanModalHeader = props => {
  const childCount = React.Children.count(props.children);
  const { title } = useModalContext();
  // Render the provided children, or default markup if none given
  return (
    <Modal.Header closeButton {...props}>
      {props.children}
      {childCount === 0 && title && <Modal.Title>{title}</Modal.Title>}
    </Modal.Header>
  );
};

ForemanModalHeader.propTypes = {
  children: PropTypes.node,
};

ForemanModalHeader.defaultProps = {
  children: null,
};

export default ForemanModalHeader;
