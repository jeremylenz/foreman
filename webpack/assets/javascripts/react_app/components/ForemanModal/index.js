import React from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { selectIsModalOpen } from './ForemanModalSelectors';
import { setModalClosed } from './ForemanModalActions';
import ForemanModal from './ForemanModal';
import ForemanModalHeader from './subcomponents/ForemanModalHeader';
import ForemanModalFooter from './subcomponents/ForemanModalFooter';
import reducer from './ForemanModalReducer';

export const reducers = { foremanModals: reducer };

const ConnectedForemanModal = props => {
  const { id, title } = props;
  const isOpen = useSelector(state => selectIsModalOpen(state, id));
  const dispatch = useDispatch();
  const onClose = () => dispatch(setModalClosed({ id }));
  return (
    <ForemanModal
      {...props}
      id={id}
      title={title}
      isOpen={isOpen}
      onClose={onClose}
    />
  );
};

ConnectedForemanModal.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string,
};

ConnectedForemanModal.defaultProps = {
  title: '',
};

// Header and Footer use the provided children, or default markup if none provided

ConnectedForemanModal.Header = ForemanModalHeader;
ConnectedForemanModal.Footer = ForemanModalFooter;

export default ConnectedForemanModal;
