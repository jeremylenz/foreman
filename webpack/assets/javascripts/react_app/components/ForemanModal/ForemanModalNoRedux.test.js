import React from 'react';
import PropTypes from 'prop-types';
import ForemanModal from '.';
import ForemanModalHeader from './subcomponents/ForemanModalHeader';
import ForemanModalFooter from './subcomponents/ForemanModalFooter';
import { testComponentSnapshotsWithFixtures } from '../../common/testHelpers';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(fn =>
    fn({
      // call the selector with this fake state
      foremanModals: {
        render: { open: true },
        closed: { open: false },
        withSuppliedChildren: { open: true },
        unordered: { open: true },
      },
    })
  ),
  useDispatch: jest.fn(action => action),
}));

const headerChild = (
  <ForemanModal.Header>this is the header</ForemanModal.Header>
);
const footerChild = (
  <ForemanModal.Footer>this is the footer</ForemanModal.Footer>
);
const modalBody = <div>This is the body</div>;

const fixtures = {
  'should render': {
    id: 'render',
    title: 'Test modal',
  },
  'should render closed': {
    id: 'closed',
    title: 'Test modal',
  },
  'renders when header and footer are supplied': {
    id: 'withSuppliedChildren',
    title: 'Test modal',
    children: [headerChild, modalBody, footerChild],
  },
  'sets open state to true based on redux state': {
    id: 'render',
    title: 'open modal',
  },
  'sets open state to false based on redux state': {
    id: 'closed',
    title: 'open modal',
  },
};

describe('ForemanModal', () => {
  describe('rendering', () => {
    testComponentSnapshotsWithFixtures(ForemanModal, fixtures);
  });
  describe('subcomponents', () => {
    it('has a Header subcomponent', () => {
      expect(ForemanModal.Header).toEqual(ForemanModalHeader);
    });
    it('has a Footer subcomponent', () => {
      expect(ForemanModal.Footer).toEqual(ForemanModalFooter);
    });
  });
  describe('PropTypes', () => {
    it('requires an id prop', () => {
      const { propTypes } = ForemanModal;
      const propsWithoutId = {
        title: 'Test modal',
      };
      PropTypes.resetWarningCache();
      expect(() =>
        PropTypes.checkPropTypes(
          propTypes,
          propsWithoutId,
          'prop',
          'ForemanModal'
        )
      ).toThrow('id');
    });
  });
});
