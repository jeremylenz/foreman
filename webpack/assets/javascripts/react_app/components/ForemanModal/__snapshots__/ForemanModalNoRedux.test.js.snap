// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForemanModal rendering renders when header and footer are supplied 1`] = `
<ForemanModal
  id="withSuppliedChildren"
  onClose={[Function]}
  openState={
    Object {
      "open": true,
    }
  }
  title="Test modal"
>
  <ForemanModalHeader>
    this is the header
  </ForemanModalHeader>
  <div>
    This is the body
  </div>
  <ForemanModalFooter>
    this is the footer
  </ForemanModalFooter>
</ForemanModal>
`;

exports[`ForemanModal rendering sets open state to false based on redux state 1`] = `
<ForemanModal
  id="closed"
  onClose={[Function]}
  openState={
    Object {
      "open": false,
    }
  }
  title="open modal"
/>
`;

exports[`ForemanModal rendering sets open state to true based on redux state 1`] = `
<ForemanModal
  id="render"
  onClose={[Function]}
  openState={
    Object {
      "open": true,
    }
  }
  title="open modal"
/>
`;

exports[`ForemanModal rendering should render 1`] = `
<ForemanModal
  id="render"
  onClose={[Function]}
  openState={
    Object {
      "open": true,
    }
  }
  title="Test modal"
/>
`;

exports[`ForemanModal rendering should render closed 1`] = `
<ForemanModal
  id="closed"
  onClose={[Function]}
  openState={
    Object {
      "open": false,
    }
  }
  title="Test modal"
/>
`;
